tests/virt_autotest/sriov_network_card_pci_passthrough.pm

```diff
@@ -215,48 +218,63 @@ sub prepare_guest {
+
+        if (is_sle('>=15-SP2') && is_kvm_host) {
+            #for sles15sp1+, PCIe replaces PCI. We need add pcie controllers to support hotplugging more SR-IOV Ethernet vf devices
+            my $controller_xml = "<controller type='pci' model='pcie-root-port'/>";
+            assert_script_run "sed -i \"/<devices>/a\\$controller_xml\\n$controller_xml\\n$controller_xml\" $vm.xml";
         }
 
-        #enable udev debug logs
-        my $udev_conf_file = "/etc/udev/udev.conf";
-        if (script_run("ssh root\@$vm \"ls $udev_conf_file\"") == 0) {
-            script_run "ssh root\@$vm \"sed -i '/udev_log *=/{h;s/^[# ]*udev_log *=.*\$/udev_log=debug/};\${x;/^\$/{s//udev_log=debug/;H};x}' $udev_conf_file";
+        elsif (is_xen_host) {
+            #disable memory ballooning for fv guest as it is not supported
+            if (is_fv_guest($vm) && $is_ballooned) {
+                assert_script_run "sed -i '/<currentMemory/d' $vm.xml";
+                record_info "Disable guest ballooning", "$vm";
+            }
+            #enable pci-passthrough and set e820_host for pv guest
+            #enable pci-passthrough for fv guest on sles15sp2+
+            #refer to bug #1167217 for the reason
+            #but need wait libvirt support for pv guest requested in a bug on sles12sp5
+            if (is_pv_guest($vm) || (is_fv_guest($vm) && is_sle('>15-SP2'))) {
+                my $passthru_xml = "<passthrough state='on'/>";
+                my $e820_xml     = "";
+                if (is_pv_guest($vm)) {
+                    $e820_xml = "<e820_host state='on'/>";
+                }
+                if (script_run("grep '<features>' $vm.xml") == 0) {
+                    assert_script_run "sed -i \"/<features>/a\\<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\" $vm.xml";
+                }
+                else {
+                    assert_script_run "sed -i \"/<domain /a\\<features>\\n<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\\n</features>\" $vm.xml";
+                }
+            }
         }
 
+        assert_script_run "virsh define $vm.xml";
+        assert_script_run "virsh start $vm";
+        sleep 60;
     }
 
-    #add pcie controllers to support hotplugging more SR-IOV Ethernet vf devices
-    my $controller_xml = "<controller type='pci' model='pcie-root-port'/>";
-    assert_script_run "sed -i \"/<devices>/a\\$controller_xml\\n$controller_xml\\n$controller_xml\" $vm.xml";
-    assert_script_run "virsh define $vm.xml";
-    assert_script_run "virsh start $vm";
-
     #passwordless access to guest
     save_guest_ip($vm, name => "br123");    #get the guest ip via key words in 'virsh domiflist'
 
+    #enable udev debug logs
+    my $udev_conf_file = "/etc/udev/udev.conf";
+    if (script_run("ssh root\@$vm \"ls $udev_conf_file\"") == 0) {
+        script_run "ssh root\@$vm \"sed -i '/udev_log *=/{h;s/^[# ]*udev_log *=.*\\\$/udev_log=debug/};\\\${x;/^\\\$/{s//udev_log=debug/;H};x}' $udev_conf_file\"";
+    }
```

Comment:
I recommend you add such ssh options:
-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no

Reply:
Makes no sense. This should be handled right after creating such guest by calling `ssh_copy_id`.

Reply:
Pavel is correct. both setup_dns and ssh_copy_id are run before the sriov tests. The host is able to access the guest passwordlessly.

Reply:
StrictHostKeyChecking means if remote host(here mean guest) changes its digest(perhaps I think), the local host(here mean host) can still connect to it passsordlessly. It loosens the known host checking on local host. I think it is useful.

Reply:
The idea of @waynechen55 is a good one. However instead of adding the `-o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no` to each ssh call, I'd suggest to add them to the `~/.ssh/config` file, e.g.

    HOST sles15sp2
        UserKnownHostsFile /dev/null
        StrictHostKeyChecking no
        User root

Like this there is only one place where the is set (in the setup) and all future `ssh` calls can profit from this. Otherwise we would need to or write a ssh wrapper function. Adding custom options to each `ssh` call renders the code unmaintainable IMHO.

Reply:
I think
HOST *
    UserKnownHostsFile /dev/null
    StrictHostKeyChecking no
    User root
also applies. Much easier.

Reply:
```
HOST *
UserKnownHostsFile /dev/null
StrictHostKeyChecking no
User root
PreferredAuthentications publickey
ControlMaster auto
ControlPersist 86400
ControlPath ~/.ssh/ssh_%r_%h_%p
```
I challenge you guys to add more options! :)

Reply:
Just note that this can be applied only after we have `authorized_keys` on each guest.

Reply:
this can be thought about in setup_dns step before testing guests rather than SRIOV tests. we will consider it in the subsequent plan.